### PR TITLE
shell: Allow users to load keys outside of home dir to ssh agent

### PR DESCRIPTION
**Pull request by petervo**

### CODE REVIEW

File: pkg/users/local.js [hunk 1]
```diff
@@ -92,7 +92,7 @@ function passwd_self(old_pass, new_pass) {
             dfd.resolve();
         })
         .fail(function(ex) {
-            if (ex.constructor.name == "ProcessError")
+            if (ex.constructor.name == "ProcessError" || ex.process_error)
```
Comment (stefwalter) [inline pkg/users/local.js]: Why do we need this at all? We should probably just check for the presence of exit_status?

File: src/base1/cockpit.js [hunk 2]
```diff
@@ -2665,6 +2665,7 @@ function factory() {
      */
 
     function ProcessError(options, name) {
+        this.process_error = true;
```
Comment (stefwalter) [inline src/base1/cockpit.js]: Ditto, don't think we need this.

File: pkg/lib/cockpit-components-file-autocomplete.jsx [hunk 3]
```diff
@@ -67,6 +67,30 @@ var FileAutoComplete = React.createClass({
         });
     },
 
+    onMouseDown: function (ev) {
```
Comment (stefwalter) [inline pkg/lib/cockpit-components-file-autocomplete.jsx]: Should this be a separate commit?

File: pkg/lib/credentials-ssh-private-keys.sh [hunk 4]
```diff
@@ -1,9 +1,25 @@
 #!/bin/sh
 
+# The first thing we do is list loaded keys
+loaded=$(ssh-add -L)
+result="$?"
+
 set -eu
```
Comment (stefwalter) [inline pkg/lib/credentials-ssh-private-keys.sh]: Why isn't the above also with -u flag?

File: pkg/lib/credentials.js [hunk 5]
```diff
@@ -288,10 +295,16 @@
...
 
-        self.unload = function unload(name) {
-            return cockpit.spawn(["ssh-add", "-d", name],
-                    { pty: true, err: "message", directory: self.path })
-                .done(refresh);
+        self.unload = function unload(key) {
+            var proc;
+            var options = { pty: true, err: "message", directory: self.path };
+
+            if (key.name && !key.agent_only)
```
Comment (stefwalter) [inline pkg/lib/credentials.js]: What's the difference between agent_only and !agent_only? Why doesn't everything go via the agent?

File: pkg/shell/credentials.js [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+                            row.remove();
+                        }
+                    }
+                });
+            });
+    }
+
+    module.exports = {
+        setup: setup
+    };
```
Comment (stefwalter) [inline pkg/shell/credentials.js]: Assuming this file is only moved code.

File: pkg/shell/index.html [hunk 7]
```diff
@@ -180,10 +180,65 @@ <h4 class="modal-title" translatable="yes">Authentication</h4>
...
+                                        <span translate>Add key</span>
+                                    </a>
+                                </td>
+                            </tr>
+                            <tr class="load-custom-key hidden">
+                                <td colspan="3">
+                                    <div class="input-group">
+                                        <div id="ssh-file-container"></div>
+                                        <span class="input-group-btn">
+                                            <button class="btn btn-default" translatable="yes">Add</button>
```
Comment (stefwalter) [inline pkg/shell/index.html]: Lets just use 'translate' it's cleaner.

File: pkg/shell/index.html [hunk 8]
```diff
@@ -180,10 +180,65 @@ <h4 class="modal-title" translatable="yes">Authentication</h4>
...
+                            <tr class="listing-ct-body">
+                                <td colspan="3">
+                                    <div class="alert alert-danger">
+                                        <i class="fa fa-exclamation-triangle"></i>
+                                        <span class="credential-alert"></span>
+                                    </div>
+                                    <div class="credential-unlock">
+                                        <dl>
+                                            <dt>
+                                                <label translatable="yes">Password</label>
```
Comment (stefwalter) [inline pkg/shell/index.html]: Ditto 'translate' attribute.

File: pkg/shell/shell.less [hunk 9]
```diff
@@ -560,6 +560,40 @@ table.credential-listing {
     width: 100%;
 }
 
+.credential-listing thead tr.load-custom-key td button {
```
Comment (stefwalter) [inline pkg/shell/shell.less]: Lets use real less nesting here.

File: pkg/shell/shell.less [hunk 10]
```diff
@@ -560,6 +560,40 @@ table.credential-listing {
...
+
+.credential-listing thead tr.load-custom-key td div.dialog-error {
+    display: none;
+}
+
+tbody.ssh-add-key-body {
+    display: none;
+}
+
+tbody.ssh-add-key-body.unlock {
```
Comment (stefwalter) [inline pkg/shell/shell.less]: Ditto about nesting.

File: pkg/shell/shell.less [hunk 11]
```diff
@@ -560,6 +560,40 @@ table.credential-listing {
...
+    display: table-row-group;
+}
+
+.credential-listing thead tr.load-custom-key td.has-error div.dialog-error {
+    display: block;
+}
```
Comment (stefwalter) [inline pkg/shell/shell.less]: Ditto about nesting.

File: test/verify/check-keys [hunk 12]
```diff
@@ -280,5 +280,86 @@ class TestKeys(MachineCase):
                 self.assertIn(old_rsa, keys)
                 self.assertIn(old_dsa, keys)
 
+        # Rhel 7.4 has an old shell
+        if m.image == 'rhel-7.4':
```
[the remark this answers is not in the document]
Comment (petervo) [reply test/verify/check-keys]: Nothing changes, we don't install the new shell during these tests. So the shell is unchanged so the test doesn't change. This code is only in the shell. The rest of the test still runs and applies

[the remark this answers is not in the document]
Comment (stefwalter) [reply test/verify/check-keys]: Aha ... i read that as RHEL 7.4 has an old "bash" shell. ... but an old cockpit shell makes sense now.

File: pkg/lib/credentials.js (hunk 5 above)
Comment (petervo) [reply pkg/lib/credentials.js]: We only know about the key because it's in the agent not in the users home dir. These are keys that we might not have corresponding .pub file for so we remove them from the agent differently.

File: pkg/shell/credentials.js (hunk 6 above)
Comment (petervo) [reply pkg/shell/credentials.js]: Moved and changed. This code only is used in the shell and since i was bringing in new deps I didn't want to bloat the shared part unnecessarily.

File: pkg/lib/cockpit-components-file-autocomplete.jsx (hunk 3 above)
Comment (petervo) [reply pkg/lib/cockpit-components-file-autocomplete.jsx]: seperated

**Comment by petervo** [comment]
Generally I don't like doing the video till after review, but hopefully it won't change to much.

https://cockpit.fedorapeople.org/add-keys.webm
